### ISSUE TITLE
[BUG] Post-import hooks in `configure_autologging_for_evaluation` configure the wrong flavor, silently disabling tracing during `mlflow.evaluate()`

### ISSUE THREAD

**Issue by Koustav-github** [mlflow/mlflow#25073]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### MLflow version

- Client: 3.15.2.dev0 (master, commit e6a3c7130936493adc148e111cfae0e9c0e712de)
- Tracking server: n/a (bug is client-side)

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 11 (10.0.26200)
- **Python version**: 3.10.18

### Describe the problem

`configure_autologging_for_evaluation` wraps every `mlflow.evaluate()`,
`mlflow.genai.evaluate()` and `optimize_prompts()` call. For each flavor it registers
a **post-import hook** that lazily enables tracing when the user's code imports that
flavor's library.

The hook closes over the loop variable `flavor`:

```python
for flavor in FLAVOR_TO_MODULE_NAME:
    original_config = global_config_snapshot.get(flavor, {}).copy()

    @autologging_conf_lock
    def _setup_autolog(module):
        autolog = get_autolog_function(flavor)     # <-- free variable
        ...                     original_config    # <-- free variable

    register_post_import_hook(_setup_autolog, FLAVOR_TO_MODULE_NAME[flavor], overwrite=True)
```

Python closures capture the *variable*, not its value, so all 17 hooks share a single
cell for `flavor`. `register_post_import_hook` fires a hook immediately **only** when
the module is already in `sys.modules`
([import_hooks/__init__.py#L104-L111](https://github.com/mlflow/mlflow/blob/master/mlflow/utils/import_hooks/__init__.py#L104-L111));
otherwise it defers it. By the time a deferred hook runs, the loop has finished and
`flavor` holds its final value — `bedrock`.

Deferring is not an edge case, it is the stated design
([trace.py#L55-L58](https://github.com/mlflow/mlflow/blob/master/mlflow/models/evaluation/utils/trace.py#L55-L58)):

> *"Using post-import hook to configure the autologging lazily when the target flavor's
> module is imported, rather than configuring it immediately. This is because the
> evaluation code usually only uses a subset of the supported flavors, hence we want to
> avoid unnecessary overhead of configuring all flavors."*

**Expected behavior:** importing `langchain` during evaluation enables LangChain trace
autologging.

**Actual behavior:** importing `langchain` configures **bedrock** autologging.
LangChain tracing is never enabled. The same holds for `openai`, `dspy`, `llama_index`
and every other flavor except whichever one sorts last in `FLAVOR_TO_MODULE_NAME`.

The failure is silent twice over: no warning is emitted when the wrong flavor is
configured, and when the wrong flavor's library is not installed the resulting
`ImportError` is swallowed by the bare `except Exception` at
[trace.py#L85](https://github.com/mlflow/mlflow/blob/master/mlflow/models/evaluation/utils/trace.py#L85)
into a `_logger.debug`. The only user-visible symptom is *"my evaluation runs have no
traces."*

`ruff --select B023` independently flags lines 62, 65, 67 and 86.

### Secondary defect in the same function

The teardown at
[trace.py#L100-L107](https://github.com/mlflow/mlflow/blob/master/mlflow/models/evaluation/utils/trace.py#L100-L107)
tests the wrong dict:

```python
for module, hooks in new_import_hooks.items():
    if hooks is None:                        # never true
        _post_import_hooks.pop(module, None)
    else:
        _post_import_hooks[module] = original_import_hooks[module]   # may write None
```

`hooks` iterates the values of `new_import_hooks`, which are always the `_setup_autolog`
function, so the `is None` branch is dead and the else-branch writes `None` into
`_post_import_hooks` for every module that had no prior hook. The condition should test
`original_import_hooks[module]`.

Because `ImportHookFinder.find_spec` gates on key *presence*
([import_hooks/__init__.py#L308](https://github.com/mlflow/mlflow/blob/master/mlflow/utils/import_hooks/__init__.py#L308)),
a `None` value keeps MLflow intercepting imports of those modules for the remaining life
of the process — a redundant `find_spec` round-trip plus an `_ImportHookChainedLoader`
wrap on every subsequent import, after every `evaluate()` call. Section **C** of the
repro below demonstrates it.

Happy to fix both in one PR, or split them — whichever you prefer.

### Tracking information

```shell
N/A - not a tracking bug. Reproduces with any tracking URI.
```

### Code to reproduce issue

```python
"""Deferred post-import hooks configure the wrong flavor. No mocking, no ML libs needed."""

import inspect
import types

from mlflow.ml_package_versions import FLAVOR_TO_MODULE_NAME
from mlflow.models.evaluation.utils.trace import configure_autologging_for_evaluation
from mlflow.utils.autologging_utils import AUTOLOGGING_INTEGRATIONS
from mlflow.utils.import_hooks import _post_import_hooks

# Flavors whose library is typically NOT yet imported -> their hooks stay deferred,
# which is the path the lazy-configuration design exists to serve.
PROBES = ["langchain", "openai", "llama_index"]

print(f"Last flavor in FLAVOR_TO_MODULE_NAME: {list(FLAVOR_TO_MODULE_NAME)[-1]!r}\n")


def flavor_bound_to(hook):
    """Report which flavor a registered hook will actually configure."""
    fn = inspect.unwrap(hook)
    bound = inspect.signature(fn).parameters.get("flavor")
    if bound is not None and bound.default is not inspect.Parameter.empty:
        return bound.default
    return inspect.getclosurevars(fn).nonlocals.get("flavor")


print("A) Which flavor does each module's deferred hook actually configure?")
with configure_autologging_for_evaluation(enable_tracing=True):
    for flavor in PROBES:
        module = FLAVOR_TO_MODULE_NAME[flavor]
        for hook in _post_import_hooks.get(module) or []:
            actual = flavor_bound_to(hook)
            verdict = "OK" if actual == flavor else "WRONG"
            print(f"  [{verdict:5}] hook for module {module!r:20} -> configures {actual!r}")

print("\nB) Fire ONLY langchain's hook. Which flavor gets configured?")
before = set(AUTOLOGGING_INTEGRATIONS)
with configure_autologging_for_evaluation(enable_tracing=True):
    module = FLAVOR_TO_MODULE_NAME["langchain"]
    for hook in _post_import_hooks.get(module) or []:
        hook(types.ModuleType(module))
    touched = set(AUTOLOGGING_INTEGRATIONS) - before
    print(f"  importing {module!r} configured autologging for: {sorted(touched) or '(none)'}")
    print("  expected: ['langchain']")

print("\nC) Teardown residue in _post_import_hooks")
leaked = {
    FLAVOR_TO_MODULE_NAME[f]: _post_import_hooks.get(FLAVOR_TO_MODULE_NAME[f])
    for f in PROBES
    if FLAVOR_TO_MODULE_NAME[f] in _post_import_hooks
}
print(f"  keys left behind after the context manager exited: {leaked or '(none)'}")
print("  expected: (none)  -- ImportHookFinder gates on key presence")
```

### Stack trace

```
No exception is raised - this is a silent misconfiguration, which is what makes it
hard to notice in practice. The observed vs. expected output of the repro above is
included under "Other info / logs".

The one exception that does occur (when the wrongly-selected flavor's library is not
installed) is swallowed at mlflow/models/evaluation/utils/trace.py:85:

    except Exception:
        _logger.debug(f"Failed to update autologging config for {flavor}.", exc_info=True)

so it is invisible at the default log level.
```

### Other info / logs

```
$ python repro.py

Last flavor in FLAVOR_TO_MODULE_NAME: 'bedrock'

A) Which flavor does each module's deferred hook actually configure?
  [WRONG] hook for module 'langchain'            -> configures 'bedrock'
  [WRONG] hook for module 'openai'               -> configures 'bedrock'
  [WRONG] hook for module 'llama_index.core'     -> configures 'bedrock'

B) Fire ONLY langchain's hook. Which flavor gets configured?
  importing 'langchain' configured autologging for: ['bedrock', 'spark']
  expected: ['langchain']

C) Teardown residue in _post_import_hooks
  keys left behind after the context manager exited: {'langchain': None, 'openai': None, 'llama_index.core': None}
  expected: (none)  -- ImportHookFinder gates on key presence
```

Expected output (what you get once `flavor` / `original_config` are bound per-iteration
and the teardown checks the right dict):

```
A) Which flavor does each module's deferred hook actually configure?
  [OK   ] hook for module 'langchain'            -> configures 'langchain'
  [OK   ] hook for module 'openai'               -> configures 'openai'
  [OK   ] hook for module 'llama_index.core'     -> configures 'llama_index'

B) Fire ONLY langchain's hook. Which flavor gets configured?
  importing 'langchain' configured autologging for: ['langchain', 'spark']

C) Teardown residue in _post_import_hooks
  keys left behind after the context manager exited: (none)
```

Note: `'spark'` appears in both runs and is unrelated - `mlflow.langchain.autolog()`
also enables the Spark datasource integration. The signal is that `langchain` is
absent from the "actual" list and present in the "expected" one.

There are currently no tests referencing `configure_autologging_for_evaluation`.

### What component(s) does this bug affect?

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [x] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

**Comment by Koustav-github** [comment]
I'd like to work on this one.

I have a patch ready that binds flavor and original_config per-iteration so each deferred hook configures its own flavor, plus unit tests covering both bindings — this function currently has no test coverage, so the tests fail on master and pass with the fix.

Two things I'd like a maintainer's steer on before I open the PR:

Scope — should the secondary teardown defect (the if hooks is None branch testing the wrong dict) go in the same PR, or a separate one?
Blast radius — the fix revives a code path that was effectively dead. For flavors where log_traces isn't supported (e.g. transformers), the hook previously resolved to bedrock and so never ran autolog(disable=True) on the right flavor. With the fix it does, which matches the docstring's stated intent but is a real behaviour change. Happy to flag anything it turns up.
Could you triage and apply the ready label when you get a chance? Per the issue template I'll hold off on the PR until then.

**Comment by PattaraS** [comment]
Thanks @Koustav-github — triaged, `ready`, and go ahead.

**Reproduced on master.** Driving `configure_autologging_for_evaluation` and firing the *deferred* hooks, importing `langchain` / `openai` / `dspy` / `llama_index` each resolves to `get_autolog_function('bedrock')` (the last `FLAVOR_TO_MODULE_NAME` key) instead of the imported flavor — so that flavor's tracing is never enabled. `ruff --select B023` independently flags the closure over `flavor`. Your secondary teardown diagnosis also checks out: the `finally` loop does `for module, hooks in new_import_hooks.items(): if hooks is None: ...`, but `new_import_hooks[module]` is always the `_setup_autolog` closure (never `None`), so it should be testing the *original* hooks, not the new ones.

Steer on your two questions:

1. **Scope** — put both fixes (per-iteration binding of `flavor`/`original_config` + the teardown dict) in **one PR, as separate commits**. Same function, same hook lifecycle, and you already have the patch plus the first test coverage this function has had; splitting the ~2-line teardown fix out is overhead with no benefit.

2. **Blast radius** — treat the revived-path behavior change as the **intended** fix (it matches the docstring; the previous state was silently-wrong-for-all-but-`bedrock`). Two asks so we can land it confidently:
   - Enumerate the per-flavor before→after matrix (before = `bedrock`, after = own flavor) and flag any flavor where the now-correct `autolog(disable=True)` could regress something relied upon.
   - Add tests for both the enable and the disable paths, including a non-tracing flavor (e.g. `transformers`), and confirm the existing `_should_enable_tracing` gate still prevents enabling tracing for flavors that shouldn't be traced during eval.

Flag anything the revived path turns up.

🤖 *Claude-assisted triage*

**Comment by Koustav-github** [comment]
Thanks for confirming the scope and guidance @PattaraS. I will combine both
fixes into a single PR with separate commits as suggested. I'll have the PR
open and ready for review by tomorrow.

**Comment by Koustav-github** [comment]
Thanks @PattaraS — both addressed.

**1. Scope** — done, one PR with the two fixes as separate commits. Each is independently green (11 tests at the first, 13 at the second) so `git bisect` stays useful.

**2. Blast radius** — while building the matrix I found the change is narrower than I'd assumed: **the immediate-fire path was never broken.** When the module is already in `sys.modules`, `register_post_import_hook` fires the hook *inside* the loop iteration, where `flavor` still holds the right value. Only the deferred path was wrong. I verified this on both patched and unpatched `trace.py`.

Per-flavor, for the deferred path only:

| Flavors | Before | After |
|---|---|---|
| `sklearn`, `pytorch`, `xgboost`, `lightgbm`, `statsmodels`, `spark`, `paddle`, `transformers` | configured `bedrock`; own flavor untouched | `autolog(disable=True)` — matches the docstring |
| 16 traceable (`langchain`, `openai`, `dspy`, `llama_index`, `anthropic`, `gemini`, `crewai`, …) | configured `bedrock` | tracing correctly enabled ← the fix |
| `pytorch-lightning`, `keras`, `tensorflow`, `haystack`, `semantic_kernel` | `ImportError`, swallowed to debug | unchanged; only the logged flavor name differs |
| `pyspark.ml` | never registered (`is_autolog_supported` is False) | unchanged |

The only genuinely new behaviour is `autolog(disable=True)` for the non-traceable flavors in row 1, and only when the library is imported *during* evaluation. In the common case it was imported before `evaluate()` was called — you needed it to build or load the model — so the hook fired immediately and `disable=True` already ran on `master`. No flavor where I'd flag a regression risk.

**Tests** — 13 total, 4 failing on `master`. Both paths covered as you asked: the enable path (`log_traces=True, log_models=False, silent=True`), the disable path (`autolog(disable=True)`), and the `_should_enable_tracing` gate parametrized over real flavors with no mocking — `langchain`/`openai`/`dspy` → `True`, `sklearn`/`xgboost`/`transformers` → `False`. Explicit opt-out (flavor-level `log_traces: False`, global `mlflow.disable`) is covered too. The two fake flavors in the hook tests are given different traceability so those tests also fail if the binding regresses.

Also ran `pytest tests/evaluate` on `master` and on the branch with identical flags — the failing set is byte-identical, no test changed status.

PR raised.

**Comment by Koustav-github** [comment]
hey @PattaraS,

just checking in, it has been a while since i raised the pr, whenever u get time, kindly review the pr, #25135

**Comment by Koustav-github** [comment]
Thanks @PattaraS for merging the pr. Happy to contribute to MLflow. :)))))))
